azure_functions_worker/extension.py

```diff
@@ -0,0 +1,297 @@
+        )
+
+        # Get application hooks from azure.functions.extension.ExtensionMeta
+        # The reutnr type is AppExtensionHooks
+        apps = sdk.ExtensionMeta.get_application_hooks()
+
+        # Invoke application hook
+        cls._safe_execute_function_load_hooks(
+            apps, APP_EXT_POST_FUNCTION_LOAD, func_name, func_directory
+        )
+
+    @classmethod
+    @enable_feature_by(
+        flag=PYTHON_ENABLE_WORKER_EXTENSIONS,
+        flag_default=(
+            PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT_39 if
+            is_python_version('3.9') else
+            PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT
+        )
+    )
+    def _invocation_extension(cls, ctx, hook_name, func_args, func_ret=None):
+        """Helper to execute extensions. If one of the extension fails in the
+        extension chain, the rest of them will continue, emitting an error log
+        of an exception trace for failed extension.
+
+        Parameters
+        ----------
+        ctx: azure.functions.Context
+            Azure Functions context to be passed onto extension
+        hook_name: str
+            The exetension name to be executed (e.g. pre_invocations).
+            These are defined in azure.functions.FuncExtensionHooks.
+        """
+        sdk = cls._try_get_sdk_with_extension_enabled()
+        if sdk is None:
+            return
+
+        # Get function hooks from azure.functions.extension.ExtensionMeta
+        # The return type is FuncExtensionHooks
+        funcs = sdk.ExtensionMeta.get_function_hooks(ctx.function_name)
+
+        # Invoke function hooks
+        cls._safe_execute_invocation_hooks(
+            funcs, hook_name, ctx, func_args, func_ret
+        )
+
+        # Get application hooks from azure.functions.extension.ExtensionMeta
+        # The reutnr type is AppExtensionHooks
+        apps = sdk.ExtensionMeta.get_application_hooks()
+
+        # Invoke application hook
+        cls._safe_execute_invocation_hooks(
+            apps, hook_name, ctx, func_args, func_ret
+        )
+
+    @classmethod
+    def get_sync_invocation_wrapper(cls, ctx, func) -> Callable[[List], Any]:
+        """Get a synchronous lambda of extension wrapped function which takes
+        function parameters
+        """
+        return functools.partial(cls._raw_invocation_wrapper, ctx, func)
+
+    @classmethod
+    async def get_async_invocation_wrapper(cls, ctx, function, args) -> Any:
+        """An asynchronous coroutine for executing function with extensions
+        """
+        cls._invocation_extension(ctx, APP_EXT_PRE_INVOCATION, args)
+        cls._invocation_extension(ctx, FUNC_EXT_PRE_INVOCATION, args)
+        result = await function(**args)
+        cls._invocation_extension(ctx, FUNC_EXT_POST_INVOCATION, args, result)
+        cls._invocation_extension(ctx, APP_EXT_POST_INVOCATION, args, result)
+        return result
+
+    @staticmethod
+    def _get_sdk_from_sys_path() -> ModuleType:
```

Comment:
Curious - could this be moved to utils? This doesn't feel like something that should be part of the ExtensionsManager - this is generic.